File: code/bench/mandel.cpp
#include <cstdio>

static long long mandel_checksum(int width, int height, int max_iter) {
  long long total = 0;
  double inv_w = 1.0 / (double)width;
  double inv_h = 1.0 / (double)height;
  for (int y = 0; y < height; ++y) {
    double ci = ((double)y * inv_h) * 2.0 - 1.0;
    for (int x = 0; x < width; ++x) {
      double cr = ((double)x * inv_w) * 3.5 - 2.5;
      double zr = 0.0;
      double zi = 0.0;
      int escaped = 0;
      int it = 0;
      for (int i = 0; i < max_iter; ++i) {
        if (!escaped) {
          double zr2 = zr * zr - zi * zi + cr;
          double zi2 = 2.0 * zr * zi + ci;
          zr = zr2;
          zi = zi2;
          if (zr * zr + zi * zi > 4.0) {
            escaped = 1;
            it = i;
          }
        }
      }
      if (!escaped) it = max_iter;
      total += it;
    }
  }
  return total;
}

int main() {
  int width = 1000;
  int height = 1000;
  int max_iter = 500;
  long long result = mandel_checksum(width, height, max_iter);
  std::printf("%lld\n", result);
  return 0;
}

File: code/bench/mandel.rs
fn mandel_checksum(width: i32, height: i32, max_iter: i32) -> i64 {
    let mut total: i64 = 0;
    let inv_w = 1.0f64 / width as f64;
    let inv_h = 1.0f64 / height as f64;
    for y in 0..height {
        let ci = (y as f64 * inv_h) * 2.0 - 1.0;
        for x in 0..width {
            let cr = (x as f64 * inv_w) * 3.5 - 2.5;
            let mut zr = 0.0_f64;
            let mut zi = 0.0_f64;
            let mut escaped = false;
            let mut it: i32 = 0;
            for i in 0..max_iter {
                if !escaped {
                    let zr2 = zr * zr - zi * zi + cr;
                    let zi2 = 2.0 * zr * zi + ci;
                    zr = zr2;
                    zi = zi2;
                    if zr * zr + zi * zi > 4.0 {
                        escaped = true;
                        it = i;
                    }
                }
            }
            if !escaped {
                it = max_iter;
            }
            total += it as i64;
        }
    }
    total
}

fn main() {
    let width = 1000;
    let height = 1000;
    let max_iter = 500;
    let result = mandel_checksum(width, height, max_iter);
    println!("{}", result);
}
